src/storage/src/table/state_table.rs

```diff
@@ -102,28 +112,124 @@ impl<S: StateStore> StateTable<S> {
         Ok(())
     }
 
-    pub async fn iter(&self, _pk: Row) -> StorageResult<StateTableRowIter<S>> {
-        todo!()
+    pub async fn iter(&self, epoch: u64) -> StorageResult<StateTableRowIter<S>> {
+        let mem_table_vec = self
+            .mem_table
+            .buffer
+            .iter()
+            .map(|(row, row_op)| (row.clone(), row_op.clone()))
+            .collect_vec();
+        StateTableRowIter::new(
+            self.keyspace.clone(),
+            self.column_descs.clone(),
```

Comment:
This seems way too heavy to create an iterator... Do we need some MVCC structures here? cc @skyzh @st1page

Reply:
we can assume the mem_table size is not very big now and try to optimize it later.

Reply:
Agree. I think we'd find a better way to implement this. I don't think this iterator can be used by arbitrary executors. It's simply too heavy.

Reply:
But the first thing to address one question -- do we expect this iterator to have lifetime?

Reply:
> we can assume the mem_table size is not very big now and try to optimize it later.

Even if it's not big, consider the case for extreme state. It will create one iterator on every group key. In one epoch we expect to process ~10k items. The cost is still large.

Reply:
Then let's add a lifetime for this iterator, and store BTreeIter inside the iterator?

For batch queries, they should use CellBasedTableIter (no lifetime) instead of this StateTableIter.

Reply:
Not sure if we can write this in safe Rust 🤣 We may have a try.

Reply:
> I don't think this iterator can be used by arbitrary executors.
I think this PR is just a start. when we actually use it in the executor, it might be ok. for example, in hash join, it might construct the iter with a pk's prefix(keyspace), then the vec's size will be ok and no overhead(because the rows in memtable must be read).

Reply:
and I agree it is not acceptable for the executors like TopN

Reply:
> For batch queries, developers should use CellBasedTableIter (no lifetime) instead of this StateTableIter.
+1 for this

Reply:
IIRC, we had some discussion about that in the past and there were some `#[self_referencing]` tricks? 
> and other question here, can we assume that during the Iter's "lifetime" there is no modify on the mem_table?

and because of the assumption above, we can use RefCell wrapping the BTreeMap and use the #[self_referencing]?